google/cloud/bigtable/data/_metrics/handlers/opentelemetry.py

```diff
@@ -0,0 +1,175 @@
+
+    The OpenTelemetry metrics that are tracked are as follows:
+      - operation_latencies: latency of each client method call, over all of it's attempts.
+      - first_response_latencies: latency of receiving the first row in a ReadRows operation.
+      - attempt_latencies: latency of each client attempt RPC.
+      - retry_count: Number of additional RPCs sent after the initial attempt.
+      - server_latencies: latency recorded on the server side for each attempt.
+      - connectivity_error_count: number of attempts that failed to reach Google's network.
+      - application_blocking_latencies: the time spent waiting for the application to process the next response.
+      - throttling_latencies: latency introduced by waiting when there are too many outstanding requests in a bulk operation.
+    """
+
+    def __init__(
+        self,
+        *,
+        project_id: str,
+        instance_id: str,
+        table_id: str,
+        app_profile_id: str | None,
+        client_uid: str | None = None,
+        **kwargs,
+    ):
+        super().__init__()
+        # otel singleton holds shared instruments
+        self.otel = _OpenTelemetryInstrumentSingleton()
+
+        # fixed labels sent with each metric update
+        self.shared_labels = {
+            "client_name": f"python-bigtable/{bigtable_version}",
+            "client_uid": client_uid or str(uuid4()),
+        }
+        if app_profile_id:
+            self.shared_labels["app_profile"] = app_profile_id
+
+    def on_operation_complete(self, op: CompletedOperationMetric) -> None:
+        """
+        Update the metrics associated with a completed operation:
+          - operation_latencies
+          - retry_count
+        """
+        labels = {
+            "method": op.op_type.value,
+            "status": op.final_status.value,
+            "streaming": op.is_streaming,
+            **self.shared_labels,
+        }
+
+        self.otel.operation_latencies.record(op.duration, labels)
+        self.otel.retry_count.add(len(op.completed_attempts) - 1, labels)
+
+    def on_attempt_complete(
+        self, attempt: CompletedAttemptMetric, op: ActiveOperationMetric
+    ):
+        """
+        Update the metrics associated with a completed attempt:
+          - attempt_latencies
+          - first_response_latencies
+          - server_latencies
+          - connectivity_error_count
+          - application_blocking_latencies
+          - throttling_latencies
+        """
+        labels = {
+            "method": op.op_type.value,
+            "status": attempt.end_status.value,
+            "streaming": op.is_streaming,
+            **self.shared_labels,
+        }
+
+        self.otel.attempt_latencies.record(attempt.duration, labels)
+        combined_throttling = attempt.grpc_throttling_time
+        if not op.completed_attempts:
+            # add flow control latency to first attempt's throttling latency
+            combined_throttling += op.flow_throttling_time
```

Comment:
I was a little unsure how the flow control throttling should be tracked. Let me know if this looks right (adding it to the throttling_latency of the first attempt)